src/replication.c

```diff
@@ -2173,6 +2235,211 @@ istgt_lu_resize_volume(spec_t *spec, uint64_t size) {
+		 * lock on spec->rq_mtx it is good enough to read replica_id
+		 * properity.
+		 */
+		found_replica = false;
+		/* If current replica not exist in provided list then that
+		 * replica need to be removed
+		 */
+		IS_REPLICA_ID_EXIST_IN_LIST(replica->replica_id,
+		    known_replica_id_list, drf, found_replica);
+		if (!found_replica) {
+			removing_replica = replica;
+			found_in_rq_list = true;
+			break;
+		}
+	}
+	if (removing_replica == NULL) {
+		TAILQ_FOREACH(replica, &spec->non_quorum_rq, r_non_quorum_next) {
+			found_replica = false;
+			IS_REPLICA_ID_EXIST_IN_LIST(replica->replica_id,
+			    known_replica_id_list, drf, found_replica);
+			if (!found_replica) {
+				removing_replica = replica;
+				break;
+			}
+		}
+	}
+
+	/* Wait for any ongoing snapshot commands */
+	while (spec->quiesce == 1) {
+		MTX_UNLOCK(&spec->rq_mtx);
+		sleep(1);
+		MTX_LOCK(&spec->rq_mtx);
+	}
+
+	if (can_replica_remove(spec, removing_replica) == false) {
+		MTX_UNLOCK(&spec->rq_mtx);
+		REPLICA_ERRLOG("Remaining replicas are not healthy to remove replica...\n");
+		return rc;
+	}
+
+	/* If replica that needs to be removed is not connected and
+	 * if there is no change in consistency model then go and update
+	 * volume configuration without pausing IOs.
+	 */
+	if (!found_in_rq_list ||
+	    (removing_replica == NULL && new_cf == spec->consistency_factor))
+		goto update_volume_configurations;
+
+	/* Select some replica other than removing replica and
+	 * wait for few seconds if there are any ongoing IOs
+	 */
+	TAILQ_FOREACH(replica, &spec->rq, r_next)
+		if (replica != removing_replica)
+			break;
+
+	spec->quiesce = 1;
+	rc = wait_for_ongoing_ios_on_replica(spec, replica, io_wait_time);
+	if ( rc != 0 ) {
+		spec->quiesce = 0;
+		MTX_UNLOCK(&spec->rq_mtx);
+		REPLICA_ERRLOG("pausing failed..\n");
+		return rc;
+	}
+	if (removing_replica != NULL) {
+		MTX_LOCK(&removing_replica->r_mtx);
+		// Marking replica as cordon(not ready to serve IOs)
+		removing_replica->cordon = 1;
+		MTX_UNLOCK(&removing_replica->r_mtx);
+		REPLICA_LOG("Replica(%s:%lu) marked cordoned for IOs\n",
+		    removing_replica->replica_id, removing_replica->zvol_guid);
+	}
+
+update_volume_configurations:
+	spec->replication_factor = new_rf;
+	spec->consistency_factor = new_cf;
+	spec->lu->replication_factor = new_rf;
+	spec->lu->consistency_factor = new_cf;
+	spec->lu->desired_replication_factor = drf;
+	spec->desired_replication_factor = drf;
+
+#ifdef REPLICATION
+	// If removing replica exists in trusty replica list remove from it
+	TAILQ_FOREACH(trusty_replica, &spec->lu->trusty_replicas, next) {
+		found_replica = false;
+		IS_REPLICA_ID_EXIST_IN_LIST(trusty_replica->replica_id,
+		    known_replica_id_list, drf, found_replica);
+		if (!found_replica) {
+			TAILQ_REMOVE(&spec->lu->trusty_replicas,
+			    trusty_replica, next);
+			break;
+		}
+	}
+#endif
+	spec->quiesce = 0;
+
+	if (removing_replica != NULL)
+		inform_mgmt_conn(removing_replica);
```

Comment:
`inform_mgmt_conn` updates the replica parameter, so better to take a replica lock prior to this function call.

Reply:
Done